### PR TITLE
Workaround for issue 40015: nuget restore uses 'resource' instead of 'resources'

### LINKED ISSUE #40015: On dotnet publish for Mac, getting unclear error "error MSB4018: The "GenerateDepsFile" task failed unexpectedly... System.InvalidOperationException: Unrecognized AssetType 'resource' "
### Describe the bug
When trying to ```dotnet publish``` a project with dependencies, I'm getting the MSBuild exception below.
Based on looking at dotnet code, I guess that the error is probably due to some typo in the package creation, where instead of ```resources```, the word ```resource``` was used (I understood it by the code around the exception thrown here:
https://github.com/dotnet/sdk/blob/a1fdd359f85eaa8a7ebb68b4856b0dc559543a93/src/Tasks/Microsoft.NET.Build.Tasks/ResolvedFile.cs#L75C17-L76C1)

However I can't I understand from the exception and the code around where should I look for that typo in our code or even in the binary or package metadata.

Short term - please advice what exactly should I look at.
Long term - consider providing clearer error message.

### To Reproduce

Dependency tree is complex and due to the unclear error message I don't know how to minimize the code into simple repro, sorry.
On private call I'll be able to provide the faulty package and to share the environment where I reproduced the bug.

### Exceptions (if any)

(Obfuscated some private paths from the logs below)

```
/usr/local/share/dotnet/sdk/8.0.202/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Sdk.targets(284,5): error MSB4018: The "GenerateDepsFile" task failed unexpectedly. [MyProject.csproj::TargetFramework=net8.0]
/usr/local/share/dotnet/sdk/8.0.202/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Sdk.targets(284,5): error MSB4018: System.InvalidOperationException: Unrecognized AssetType 'resource' for /Volumes/.../%NugetMachineInstallRoot%/.pkgref/PackageName/2.127.7249-prerelease-007/runtimes/osx/lib/net6.0/ar/BinaryName.resources.dll [MyProject.csproj::TargetFramework=net8.0]
/usr/local/share/dotnet/sdk/8.0.202/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Sdk.targets(284,5): error MSB4018:    at Microsoft.NET.Build.Tasks.ResolvedFile..ctor(ITaskItem item, Boolean isRuntimeTarget) [MyProject.csproj::TargetFramework=net8.0]
/usr/local/share/dotnet/sdk/8.0.202/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Sdk.targets(284,5): error MSB4018:    at Microsoft.NET.Build.Tasks.GenerateDepsFile.<>c.<WriteDepsFile>b__120_3(ITaskItem f) [MyProject.csproj::TargetFramework=net8.0]
/usr/local/share/dotnet/sdk/8.0.202/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Sdk.targets(284,5): error MSB4018:    at System.Linq.Enumerable.SelectArrayIterator`2.MoveNext() [MyProject.csproj::TargetFramework=net8.0]
/usr/local/share/dotnet/sdk/8.0.202/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Sdk.targets(284,5): error MSB4018:    at System.Linq.Enumerable.ConcatIterator`1.MoveNext() [MyProject.csproj::TargetFramework=net8.0]
/usr/local/share/dotnet/sdk/8.0.202/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Sdk.targets(284,5): error MSB4018:    at System.Linq.Lookup`2.Create(IEnumerable`1 source, Func`2 keySelector, IEqualityComparer`1 comparer) [MyProject.csproj::TargetFramework=net8.0]
/usr/local/share/dotnet/sdk/8.0.202/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Sdk.targets(284,5): error MSB4018:    at System.Linq.GroupedEnumerable`2.GetEnumerator() [MyProject.csproj::TargetFramework=net8.0]
/usr/local/share/dotnet/sdk/8.0.202/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Sdk.targets(284,5): error MSB4018:    at Microsoft.NET.Build.Tasks.DependencyContextBuilder.WithResolvedNuGetFiles(IEnumerable`1 resolvedNuGetFiles) [MyProject.csproj::TargetFramework=net8.0]
/usr/local/share/dotnet/sdk/8.0.202/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Sdk.targets(284,5): error MSB4018:    at Microsoft.NET.Build.Tasks.GenerateDepsFile.WriteDepsFile(String depsFilePath) [MyProject.csproj::TargetFramework=net8.0]
/usr/local/share/dotnet/sdk/8.0.202/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Sdk.targets(284,5): error MSB4018:    at Microsoft.NET.Build.Tasks.TaskBase.Execute() [MyProject.csproj::TargetFramework=net8.0]
/usr/local/share/dotnet/sdk/8.0.202/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Sdk.targets(284,5): error MSB4018:    at Microsoft.Build.BackEnd.TaskExecutionHost.Microsoft.Build.BackEnd.ITaskExecutionHost.Execute() [MyProject.csproj::TargetFramework=net8.0]
/usr/local/share/dotnet/sdk/8.0.202/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Sdk.targets(284,5): error MSB4018:    at Microsoft.Build.BackEnd.TaskBuilder.ExecuteInstantiatedTask(ITaskExecutionHost taskExecutionHost, TaskLoggingContext taskLoggingContext, TaskHost taskHost, ItemBucket bucket, TaskExecutionMode howToExecuteTask) [MyProject.csproj::TargetFramework=net8.0]
```

### Further technical details
```
- Include the output of `dotnet --info`
.NET SDK:
 Version:           8.0.202
 Commit:            25674bb2f4
 Workload version:  8.0.200-manifests.4e94be9c

Runtime Environment:
 OS Name:     Mac OS X
 OS Version:  14.4
 OS Platform: Darwin
 RID:         osx-arm64
 Base Path:   /usr/local/share/dotnet/sdk/8.0.202/

.NET workloads installed:
There are no installed workloads to display.

Host:
  Version:      8.0.3
  Architecture: arm64
  Commit:       9f4b1f5d66

.NET SDKs installed:
  6.0.420 [/usr/local/share/dotnet/sdk]
  7.0.407 [/usr/local/share/dotnet/sdk]
  8.0.202 [/usr/local/share/dotnet/sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.App 6.0.28 [/usr/local/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 7.0.17 [/usr/local/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 8.0.3 [/usr/local/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 6.0.28 [/usr/local/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 7.0.17 [/usr/local/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 8.0.3 [/usr/local/share/dotnet/shared/Microsoft.NETCore.App]

Other architectures found:
  x64   [/usr/local/share/dotnet/x64]
    registered at [/etc/dotnet/install_location_x64]

Environment variables:
  Not set

global.json file:
  /Volumes/.../global.json

Learn more:
  https://aka.ms/dotnet/info

Download .NET:
  https://aka.ms/dotnet/download
```

- The IDE (VS / VS Code/ VS4Mac) you're running on, and its version
No IDE, running the following line from command line:
```dotnet publish --interactive -c Debug-Mac -f net8.0 <project-path>```

### PR DESCRIPTION
**Pull request by YehezkelShB**
Fixes #40015

Tested locally with a project manifesting the issue (throwing `System.InvalidOperationException: Unrecognized AssetType 'resource'`), and with the new `Microsoft.NET.Build.Tasks.dll` it doesn't happen

### CODE REVIEW

**Review by dsplaisted — APPROVED** [review]
I'm not sure about the details of how this issue happens.  It sounds like it may be a malformed package that causes the error, but the package apparently agrees with NuGet about what the asset type should be called.

Anyway, this seems like it will probably fix the issue or at least not make things worse, so I think it's a good idea.

**Comment by yitzhaks** [comment]
/ba-g Build failure is wholly unrelated to this change

**Comment by YehezkelShB** [comment]
Seems like all the checks have passed 🥳
Is it possible to merge it, please, or there is anything else I have to do?
Thanks!

**Comment by YehezkelShB** [comment]
Any chance to backport this to release/8.0.4xx and/or release/9.0.1xx? 🙏
@dsplaisted @marcpopMSFT @baronfel

**Comment by marcpopMSFT** [comment]
9.0.2xx yes. I don't know if we have enough feedback to justify taking it for 1xx or 4xx which are fairly locked down. The risk looks low but generally to get approval we need a strong customer justification.

How long has this been broken (I think the answer is always)? Why is this coming up more recently as impacting rather than in previous releases?

**Comment by YehezkelShB** [comment]
It isn't new, as you said, but only lately we started to care as we now need to build a project for new .NET and then it fails (for .Net Fx it works somehow). Adding this fix to 8.0.4xx will unblock us in a reasonable timeline. Waiting to 9.0.2xx, or even waiting for the monorepo to start using 9.0.1xx, means we have to find a workaround with today SDK.

To be clear: if there is any workaround suggestion, we'd be very happy to try it so we can be unblocked without burdening you with backporting considerations. Thanks in advance!

**Comment by YehezkelShB** [comment]
Well, seems like 8.0.4xx series isn't interesting for us anymore. Not sure if I am still pursuing a backport for 9.0.1xx either as it seems that switching from `RuntimeIdentifiers` definition in the project to using separated `RuntimeIdentifier` definitions, avoids this issue. Does it make sense to you?

**Comment by dsplaisted** [comment]
To consider backporting to 9.0.1xx, what we would need is a better description / understanding of what scenario is broken.   My understanding is that the issue is with NuGet packages that have an asset type of `resource`.  That seems to be a rare occurrence, otherwise we'd be seeing a lot more people hit this bug.  How are such packages created?  Where do they come from?

Thanks.

**Comment by YehezkelShB** [comment]
There is no `assetType: resource` added anywhere in the NuGet package creation. As we saw in the discussion in the original issue, the problem comes from NuGet restore code directly. The point is that NuGet adds assetType only if the asset doesn't have RID assigned in the NuGet creation, which makes sense for language resource DLLs, but it isn't necessarily what everyone does. Also, as mentioned above, it seems like using RuntimeIdentifier (singular) in the consumer project instead of RuntimeIdentifiers (plural) also avoids this issue (I guess the SDK doesn't look at assetType at all in this case).

Anyway, it seems like we are fine with the current state so no additional backport is needed as much as we care. Thanks!

**Pull request merged**
